pulser/simulation/simresults.py

```diff
@@ -14,63 +14,319 @@
 """Classes for containing and processing the results of a simulation."""
 
 from __future__ import annotations
-
-from typing import Optional, Union
+from collections import Counter
+from abc import ABC, abstractmethod
+from typing import Optional, Union, cast, Tuple
 from collections.abc import Sequence
 
+import matplotlib.pyplot as plt
 import qutip
+from qutip.piqs import isdiagonal
 import numpy as np
 from numpy.typing import ArrayLike
 
 
-class SimulationResults:
+class SimulationResults(ABC):
     """Results of a simulation run of a pulse sequence.
 
+    Parent class for NoisyResults and CleanResults.
     Contains methods for studying the states and extracting useful information
     from them.
     """
 
-    def __init__(self, run_output: Sequence[qutip.Qobj], dim: int, size: int,
-                 basis_name: str, meas_basis: Optional[str] = None) -> None:
+    def __init__(self, dim: int, size: int, basis_name: str,
+                 sim_times: ArrayLike) -> None:
         """Initializes a new SimulationResults instance.
 
         Args:
-            run_output (list[qutip.Qobj]): List of ``qutip.Qobj`` corresponding
-                to the states at each time step after the evolution has been
-                simulated.
             dim (int): The dimension of the local space of each atom (2 or 3).
             size (int): The number of atoms in the register.
             basis_name (str): The basis indicating the addressed atoms after
                 the pulse sequence ('ground-rydberg', 'digital' or 'all').
             meas_basis (Optional[str]): The basis in which a sampling
                 measurement is desired.
+            sim_times (array): Array of times (µs) when simulation results are
+                returned.
         """
-        self._states = run_output
         self._dim = dim
         self._size = size
         if basis_name not in {'ground-rydberg', 'digital', 'all'}:
             raise ValueError(
                 "`basis_name` must be 'ground-rydberg', 'digital' or 'all'."
                 )
         self._basis_name = basis_name
+        self.sim_times = sim_times
+        self._results: Union[list[Counter], list[qutip.Qobj]]
+        self._weights: list[float]
+
+    @abstractmethod
+    def get_state(self, t: int) -> qutip.Qobj:
+        """Returns the state of the system at time t."""
+        pass
+
+    @abstractmethod
+    def get_final_state(self) -> qutip.Qobj:
+        """Returns the final state of the system."""
+        pass
```

Comment:
The problem here is that the signature for `CleanResults.get_state` and its `NoisyResults` counterpart are different, there are more arguments in the `CleanResults` case...

Reply:
Yes... Ideally to me this would be done using `functools.partialmethod`, but you can't define the docstring cleanly with it.  You can take a look, but if it doesn't lead anywhere, the other option would be to define `get_state` here with `*args` and `**kwargs`, but I don't know if that's better. I just wanted to avoid the code duplication